jcstress-samples/src/main/java/org/openjdk/jcstress/samples/concurrency/classic/ClassicProblem_02_ProducerConsumerProblem.java

```diff
@@ -0,0 +1,312 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jcstress.samples.concurrency.classic;
+
+import org.openjdk.jcstress.annotations.*;
+import org.openjdk.jcstress.infra.results.III_Result;
+import org.openjdk.jcstress.infra.results.Z_Result;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE_INTERESTING;
+
+/*
+    How to run this test:
+        $ java -jar jcstress-samples/target/jcstress.jar -t ClassicProblem_02_ProducerConsumerProblem
+ */
+
+/**
+ * This sample shows you how JCStress can help you to test solutions for the famous producer-consumer problem.
+ * See https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem for more information about the problem and solutions.
+ */
+public abstract class ClassicProblem_02_ProducerConsumerProblem {
+    private final static int BUFFER_SIZE = 2;
+
+    static class SemaphoresBase {
+        protected final Semaphore fillCount = new Semaphore(BUFFER_SIZE);
+        protected final Semaphore emptyCount = new Semaphore(BUFFER_SIZE);
+
+        public SemaphoresBase() {
+            try {
+                fillCount.acquire(BUFFER_SIZE);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public int produce() {
+            try {
+                // produce item
+                emptyCount.acquire();
+                int index = putItemIntoBuffer();
+                fillCount.release();
+                return index;
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public int consume() {
+            try {
+                int count = count();
+                fillCount.acquire();
+                takeItemFromBuffer();
+                emptyCount.release();
+                // consume item
+                return count;
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        protected int putItemIntoBuffer() {
+            return count();
+        }
+
+        protected void takeItemFromBuffer() { }
+
+        protected int count() {
+            return fillCount.availablePermits();
+        }
+    }
+
+    @JCStressTest
+    @Outcome(id = {"true"}, expect = ACCEPTABLE)
```

Comment:
`desc`, please.

Reply:
I have added one and used it consistently in the other sub samples. I am not sure if you like it. Please feel free to tell me a better one if one comes to your mind,